music_grabber.py

```diff
@@ -0,0 +1,123 @@
+import os
+import logging
+from youtube_dl import YoutubeDL
+from urllib.parse import urlparse
+from urllib.request import Request
+from urllib.request import urlopen
+from PIL import Image
+import config
+
+DOWNLOAD_DIR = config.DOWNLOAD_DIR
+
+logger = logging.getLogger(__name__)
+
+
+class WrongCategoryError(ValueError):
+    pass
+
+
+class WrongFileSizeError(ValueError):
+    pass
+
+
+class MusicDownloader:
+
+    def download(self, input_url, filesize_limit):
+        ydl_opts = {
+                'format': 'bestaudio[protocol^=http]',
+                'outtmpl': DOWNLOAD_DIR + '%(extractor)s_%(id)s.%(ext)s',
+                'writethumbnail': True
+        }
+        ydl = YoutubeDL(ydl_opts)
+        info = ydl.extract_info(input_url, download=False)
+        webpage_info = info['extractor'] + " (" + info['id'] + ")"
+        info_format = next((item for item in info['formats']
+                            if item['format_id'] == info['format_id']),
+                           None)
+
+        if self.__youtubeVideoNotMusic(info):
+            raise WrongCategoryError(webpage_info +
+                                     ": not under Music category")
+
+        audio_filesize = self.__getFileSize(info_format)
+        if int(audio_filesize) < filesize_limit:
+            logger.info(webpage_info + ": downloading...")
+            ydl.process_info(info)
+            info = self.__addDownloadsSubset(info)
+            if self.__checkDownloadedFiles(info['downloads']):
+                return info
+            else:
+                raise ValueError(webpage_info +
+                                 ": failed to download audio or thumbnail")
+        else:
+            raise WrongFileSizeError(webpage_info + ": audio file size (" +
+                                     str(audio_filesize) + ") is greater" +
+                                     "than the limit: (" +
+                                     str(filesize_limit) + ")")
+
+    def __youtubeVideoNotMusic(self, info):
+        if info['extractor'] == 'youtube' and \
+                'Music' not in info['categories']:
+            return True
+        else:
+            return False
+
+    def __getFileSize(self, info_format):
+        if 'filesize' in info_format:
```

Comment:
just wondering, in which situations will **filesize** be in `info_format`? is it when you download from soundcloud? and for youtube i presume you check `Content-Length`?

Reply:
From what I've experienced, it's usually direct HTTP on YouTube (DASH). For SoundCloud it usually does not have `filesize` field.

youtube-dl has extractors for each sites([YouTube](https://github.com/ytdl-org/youtube-dl/blob/master/youtube_dl/extractor/youtube.py#L1823 ), [SoundCloud](https://github.com/ytdl-org/youtube-dl/blob/master/youtube_dl/extractor/soundcloud.py#L341)), seems the filesize field could be added on any of the site, just some cases it can't get the value, it's may also get by sending HTTP HEAD request, when the info_dict has filesize for this format, we just use it.

You can try this on commandline (after install youtube-dl from package manager), there are usually filesize for YouTube but not SoundCloud.
`youtube-dl -F <youtube_or_soundcloud_url>`